core/src/test/scala/unit/kafka/tools/DumpLogSegmentsTest.scala

```diff
@@ -243,6 +246,209 @@ class DumpLogSegmentsTest {
+    log.flush(false)
+
+    val expectedUpdatePayload = String.format("RemoteLogSegmentMetadataUpdate{remoteLogSegmentId=" +
+      "RemoteLogSegmentId{topicIdPartition=%s:%s-0, id=%s}, customMetadata=Optional[" +
+      "CustomMetadata{4 bytes}], state=COPY_SEGMENT_FINISHED, eventTimestampMs=0, brokerId=0}", topicId, topicName, remoteSegmentId)
+    val expectedDeletePayload = String.format("RemotePartitionDeleteMetadata{topicPartition=%s:%s-0, " +
+      "state=DELETE_PARTITION_MARKED, eventTimestampMs=0, brokerId=0}", topicId, topicName)
+    
+    val output = runDumpLogSegments(Array("--remote-log-metadata-decoder", "--files", logFilePath))
+    assertTrue(output.contains("Log starting offset: 0"))
+    assertTrue(output.contains(expectedUpdatePayload))
+    assertTrue(output.contains(expectedDeletePayload))
+  }
+  
+  @Test
+  def testDumpRemoteLogMetadataMultipleRecordsMultipleBatches(): Unit = {
+    val topicId = Uuid.randomUuid
+    val topicName = "foo"
+    val remoteSegmentId = Uuid.randomUuid
+    
+    val topicIdPartition = new TopicIdPartition(topicId, new TopicPartition(topicName, 0))
+    val remoteLogSegmentId = new RemoteLogSegmentId(topicIdPartition, remoteSegmentId)
+
+    val metadata = Seq(
+      new RemoteLogSegmentMetadataUpdate(remoteLogSegmentId, time.milliseconds,
+        Optional.of(new RemoteLogSegmentMetadata.CustomMetadata(Array[Byte](0, 1, 2, 3))), RemoteLogSegmentState.COPY_SEGMENT_FINISHED, 0),
+      new RemotePartitionDeleteMetadata(topicIdPartition, RemotePartitionDeleteState.DELETE_PARTITION_MARKED, time.milliseconds, 0)
+    )
+
+    val records: Array[SimpleRecord] = metadata.map(message => {
+      new SimpleRecord(null, new RemoteLogMetadataSerde().serialize(message))
+    }).toArray
+
+    val logConfig = LogTestUtils.createLogConfig(segmentBytes = 1024 * 1024)
+    log = LogTestUtils.createLog(logDir, logConfig, new BrokerTopicStats, time.scheduler, time)
+    log.appendAsLeader(MemoryRecords.withRecords(Compression.NONE, records:_*), leaderEpoch = 0)
+    log.appendAsLeader(MemoryRecords.withRecords(Compression.NONE, records:_*), leaderEpoch = 0)
+    log.flush(false)
+
+    val expectedUpdatePayload = String.format("RemoteLogSegmentMetadataUpdate{remoteLogSegmentId=" +
+      "RemoteLogSegmentId{topicIdPartition=%s:%s-0, id=%s}, customMetadata=Optional[" +
+      "CustomMetadata{4 bytes}], state=COPY_SEGMENT_FINISHED, eventTimestampMs=0, brokerId=0}", topicId, topicName, remoteSegmentId)
+    val expectedDeletePayload = String.format("RemotePartitionDeleteMetadata{topicPartition=%s:%s-0, " +
+      "state=DELETE_PARTITION_MARKED, eventTimestampMs=0, brokerId=0}", topicId, topicName)
+
+    val output = runDumpLogSegments(Array("--remote-log-metadata-decoder", "--files", logFilePath))
+    assertTrue(output.contains("Log starting offset: 0"))
+    assertTrue(countSubstring(output, expectedUpdatePayload) == 2)
+    assertTrue(countSubstring(output, expectedDeletePayload) == 2)
```

Comment:
I was thinking how we can make a better assertion to assert that:
1. There are indeed 2 batches, not 1, not 3
2. There are no additional records inside the output

Maybe we can add more assertions:
1. assert `baseOffset:` count, or other strings, to verify there are only 2 batches
2. assert `payload:` count is 4, or something similar, to verify there are 4 records only.
3. So, then, we can verify among these 4 records, there are 2 `expectedUpdatePayload`, and 2 `expectedDeletePayload`. 

WDYT? Same comments apply to other test cases.

Reply:
Yes, this makes sense. Thanks.